### PR TITLE
 Add an OnError handler, report the reason for disconnects

### LINKED ISSUE #21: client net.Conn in onConnect Hook
In onConnect event it would be interesting to be able to read client net.Conn. For example to view remoteAddr.

### LINKED ISSUE #33: Show the reason for client disconnects
While debugging an application, I was at some point confounded by what looked to both the client and the server like the other had side had closed the connection. On further investigation, there were two identical clients, one a zombie process, that were both trying to connect, and each time one connects the other is disconnected.

The errors that were printed in OnDisconnect both do not include the remote address field, which revealed the presence of duplicate client IDs, these also happened to look like ordinary client-initiated disconnects, so the server logs would not reveal there to be a session takeover.

Auditing the code, I learned that both an explicit Disconnect packet could trigger disconnection, but so could EOF or a variety of errors in several code paths. The existing Client was well synchronized for this, but the original cause of stopping the connection was being lost. This adds the first cause to stop the connection. In EstablishConnection(), the root cause will be the one passed to OnDisconnect, so the server logs will reveal the root of the problem. (Since this is MQTT 3.x, there is no way for the Disconnect packet to tell the client, so it could place an explanation in its own logs.)

In addition, there would be no way for a user to configure visibility into other errors, which might less obviously be the cause of or associated with disconnection or simply provide warning flags. Proposing to add an `OnError` handler to capture support logging all errors. 

Related to #21, since if there are duplicate clients it will be nice to say where they're coming from.

### PR DESCRIPTION
**Pull request by jmacd**
Redo of #34.
Fixes #33.
Fixes #21.
Incorporates #36. 

Adds an OnError handler for errors that are not reported via the OnDisconnect handler.

Results from auditing error propagation in the server.
- an error was being skipped from ResendInflight
- an error was being ignored in `Reader.ReadFrom()`, which wasn't causing harm but made diagnosing other issues (and test coverage) more difficult

Adds a "Session reestablished" error to indicate when one client connection causes another to disconnect. For example, when two clients use the same ID, the server logs might read:

```
<< OnConnect client connected operator from [::1]:56779
<< OnDisconnect client disconnected operator connection from [::1]:56779: Session reestablished
```

Extends a few existing tests to hit test coverage.  Extensive testing reveals no new issues.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -166,6 +173,46 @@ func (s *Server) inlineClient() {
...
+	// If one of these ordinary conditions stops the connection,
+	// then the client closed or broke the connection.
+	if s.Events.OnError != nil &&
+		!errors.Is(err, io.EOF) {
+		s.Events.OnError(cl, err)
+	}
+	return err
+}
+
+func (s *Server) onStorage(cl events.Clientlike, err error) {
```
Comment (mochi-co) [inline server/server.go:209]: This is a nice addition

File: server/events/events.go [hunk 2]
```diff
@@ -15,15 +15,14 @@ type Packet packets.Packet
...
 
-// FromClient returns an event client from a client.
-func FromClient(cl *clients.Client) Client {
-	return Client{
-		ID:       cl.ID,
-		Listener: cl.Listener,
-	}
+// Clientlike is an interface for Clients and client-like objects that
+// are able to describe their client/listener IDs and remote address.
+type Clientlike interface {
```
Comment (mochi-co) [inline server/events/events.go:24]: This is a good idea.

File: server/server.go [hunk 3]
```diff
@@ -378,7 +421,7 @@ func (s *Server) Publish(topic string, payload []byte, retain bool) error {
 	}
 
 	if retain {
-		s.retainMessage(pk)
+		s.retainMessage(&s.inline, pk)
```
Comment (mochi-co) [inline server/server.go:424]: I wonder if this solves https://github.com/mochi-co/mqtt/issues/37 🤔

File: server/server_test.go [hunk 4]
```diff
@@ -234,12 +276,8 @@ func TestServerEventOnConnect(t *testing.T) {
 	s, cl, _, _ := setupClient()
 	s.Clients.Add(cl)
 
-	var hookedClient events.Client
-	var hookedPacket events.Packet
-	s.Events.OnConnect = func(cl events.Client, pk events.Packet) {
-		hookedClient = cl
-		hookedPacket = pk
-	}
+	var hook packetHook
```
Comment (mochi-co) [inline server/server_test.go:279]: much tidier 👍🏻

**Review by mochi-co — APPROVED** [review]
Amazing work. I approve all these changes.

File: server/server.go (hunk 3 above)
Comment (stffabi) [reply server/server.go:424]: Unfortunately this PR didn't fix #37, it seems like the retain flag is missing in the FixedHeader. I've created PR #42 which seems to fix the retain in my case.

**Pull request merged**